### PR TITLE
snowflake gcs staging

### PR DESCRIPTION
**Pull request by jrhizor**
Uses GCS for staging to Snowflake as an alternative to S3.

### CODE REVIEW

**Review by ChristopheDuong — APPROVED** [review]

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/gcs/GcsStreamCopier.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        return destTableName;
+    }
+
+    @Override
+    public String generateMergeStatement(String destTableName) throws Exception {
+        LOGGER.info("Preparing to merge tmp table {} to dest table: {}, schema: {}, in destination.", tmpTableName, destTableName, schemaName);
+        var queries = new StringBuilder();
+        if (destSyncMode.equals(DestinationSyncMode.OVERWRITE)) {
+            queries.append(sqlOperations.truncateTableQuery(schemaName, destTableName));
+            LOGGER.info("Destination OVERWRITE mode detected. Dest table: {}, schema: {}, truncated.", destTableName, schemaName);
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/gcs/GcsStreamCopier.java]:
```suggestion
            LOGGER.info("Destination OVERWRITE mode detected. Dest table: {}, schema: {}, will be truncated.", destTableName, schemaName);
```

File: airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json [hunk 2]
```diff
@@ -145,6 +145,36 @@
...
+                "title": "GCS Bucket Name",
+                "type": "string",
+                "description": "The name of the staging GCS bucket. Airbyte will write files to this bucket and read them via <pre>COPY</pre> statements on Snowflake.",
+                "examples": ["airbyte-staging"],
+                "order": 1
+              },
+              "credentials_json": {
+                "title": "Google Application Credentials",
+                "type": "string",
+                "description": "The contents of the JSON key file that has read/write permissions to the staging GCS bucket. You will separately need to grant bucket access to your Snowflake GCP service account.",
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-snowflake/src/main/resources/spec.json:172]: we should point them to a link on how to get these creds

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/AbstractStreamCopier.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import io.airbyte.db.jdbc.JdbcDatabase;
+import io.airbyte.integrations.destination.ExtendedNameTransformer;
+import io.airbyte.integrations.destination.jdbc.SqlOperations;
+import io.airbyte.protocol.models.DestinationSyncMode;
+
+public abstract class AbstractStreamCopier implements StreamCopier {
```
Comment (cgardens) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/AbstractStreamCopier.java:53]: i'm not super convinced that this class is getting you that much. is it just trying to save the 14 lines of code of the field declarations and assignments? if there's something more to it that i'm missing then fair, but if it's just to save those lines of code, i don't think the added indirection is worth it.

**Review by cgardens — APPROVED** [review]
nice!

**Pull request merged**